model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/entities/CachedClient.java

```diff
@@ -166,12 +184,16 @@ public int getNotBefore() {
         return notBefore;
     }
 
+    public Set<String> getScope(Supplier<ClientModel> client) {
+        return scope.get(client);
+    }
+
     public Set<String> getScope() {
-        return scope;
+        return scope.get();
```

Comment:
This changes semantics from the previous version. If the data has not been loaded, then a "fallback" value would be returned that may be out of sync. This may be desired, however it needs to be done in a differently named method, to clearly state that invalid value may be returned.
@mhajas cc

Reply:
Done

Reply:
This needs to be done properly. The `getScope()` method is used e.g. from `HasRolePredicate` that would now get the default (empty) value and would stop working. Thus the semantics of `getScope` has changed in incompatible way and needs fixing. Perhaps we can leave the scope as the only non-lazily loaded parameters?

Reply:
I don't understand what you are saying. The scope was previously initialized with an empty value, no ?

Why it would stop working?

Reply:
Originally the scope has been initialized with the set of scope loaded from the database.
In this PR, this changes: if the scope has not been initialized via `getScope(Supplier<ClientModel> client)` and rather queried by this method `getScope()`, this would fall back to the `tryGetOrDefault` which, when not loaded, would return the `fallback` value (i.e. empty set).
It needs to be fixed so that the `getScope()` always returns the proper scopes.

Reply:
I see. However, when running the predicate an empty scope list should not impact invalidations because the proper scopes should be loaded/updated when some code tries to access it. That is my thinking ...

Do you have a specific scenario I can test to make sure there is an issue? My point being that loading scope mappings should really be delayed ...

Reply:
For the cached client, e.g. invalidations on a different node in the cluster. For client scope variant, see the usages below:

![image](https://user-images.githubusercontent.com/5391360/104475323-3f0f1680-55bf-11eb-9676-a104696362db.png)

There are only two options - either storing the scopes in the cached object, or always providing the supplier in all the places where `getScope()` is used from, effectively removing `getScope()` and using solely `getScope(Supplier)`.

Reply:
Humm ... client scope is not good ...

For the client, I'm still not convinced. Even when considering invalidations on other nodes.

I'll try to come up with something that does not involve pre-loading scope mappings. If it gets too complicated, I'll just fallback to pre-loading.

Makes sense?

Reply:
> For the client, I'm still not convinced. Even when considering invalidations on other nodes.

Is data consistency a convincing argument? Is working with wrong roles assigned to a client (as a result of not invalidating client when it should have been) a convincing argument?
cc: @mposolda

> I'll try to come up with something that does not involve pre-loading scope mappings. If it gets too complicated, I'll just fallback to pre-loading.

Sounds good.

Reply:
Yeah, it is. But I'm not sure if it is true when we are talking about the usage of predicates (differently than client scope). Like when for clients.

As I mentioned, even though an cached entry has the wrong roles just because it was not yet initialized, calling code (when using the proper get variant) should effectively return the correct roles.

I would love to have a test scenario for this, as it seems we don't cover this in our clustering tests (all tests are passing).  So maybe I should just be "conservative" and pre-load clients too.

Reply:
@hmlnarik I did some tests using a cluster and checking whether or not scope mappings are consistent across nodes after creating and removing roles as well as assigning/unassigning roles from the clients.

The behavior I mentioned is what I see.

I may be missing something, so I appreciate it if you can try the same. Would also be nice to have some feedback from @mposolda to confirm if that is true and if he considered this during his initial review.

Reply:
@pedroigor I did overlook that this can be a potential issue during my initial review. I did not review thoroughly as this is just draft, but anyway, I should spot this, my bad...

I agree that inconsistency can probably happen. It is probably very hard to simulate - you would need to have client in the cache on both cluster nodes, but make sure that "roleScopeMappings" for the particular client are not yet loaded. In this case, HasRolePredicate may miss to invalidate the "partially loaded" client and stale cache entry can in theory happen in the cache. I can't guarantee 100% it won't happen. Those issues are hard to simulate. Usually, you can't simulate them with the automated tests and not even manually, but they often happen just in the production (customers environment) :-)

IMO it may be better to be on the safe side and just skip lazy loading for roleScopeMappings of clientScopes and clients? We can just load roleScopeMappings eagerly. We will be safe and the performance penalty of this is really negligible due the:
- Cached ClientScopes are already lazily loaded when realm is cached due your other changes
- Cached clients are also lazily loaded just when needed
- There is not much benefit of lazy loading role scope mappings as they are usually needed in same HTTP request when the particular CachedClientScope is needed. It looks that Cached clientScope is cached during the "code-to-token" request or "refresh-token" request. But same HTTP requests also calls "CachedClientScope.getScope()" and requires roleScopeMappings.
- The client is usually cached at the point when OIDC AuthorizationEndpoint is called for this client (Login page is shown). The CachedClient.getScope() is called after successful authentication. So here there is small benefit of lazy loading as roleScopeMappings are not necessarily cached at the same HTTP request as the client itself. However quite negligible as user usually successfully finish the login once the client is first displayed.

Reply:
Safe or not safe, that is the question :)

I did exactly what you mentioned when testing, and I just couldn't spot any issue. I made sure the client in node-1 has empty scopes, whereas the same client in node-2 is fully loaded with them. I removed roles, created new ones, assigned, unassigned, etc. No success. I really want to be wrong here, but I won't spend more time on this.

I also think is easier to be safe. But for this particular case, I'm keen to take the risk and trust our test suite and the test scenario you just described.

You're right. Performance-wise, it is one less set of queries. But it is not so critical.

Changes on the way.